astropy/coordinates/representation.py

```diff
@@ -1593,6 +1618,47 @@ def represent_as(self, other_class, differential_class=None):
 
         return super().represent_as(other_class, differential_class)
 
+    def transform(self, matrix):
+        r"""Transform the unit-spherical coordinates using a 3x3 matrix.
+
+        This returns a new representation and does not modify the original one.
+        Any differentials attached to this representation will also be
+        transformed.
+
+        Parameters
+        ----------
+        matrix : (3,3) array-like
+            A 3x3 matrix, such as a rotation matrix (or a stack of matrices).
+
+        Returns
+        -------
+        `UnitSphericalRepresentation` or `SphericalRepresentation`
+            If ``matrix`` is O(3) -- :math:`M \dot M^T = I` -- like a rotation,
+            then the result is a `UnitSphericalRepresentation`.
+            All other matrices will change the distance, so the dimensional
+            representation is used instead.
+
+        """
+        # TODO! shortcut if there are differentials.
+        # Currently just super, which uses Cartesian backend.
+        if self.differentials:
+            rep = super().transform(matrix)
+
+        # the transformation matrix does not need to be a rotation matrix,
+        # so the unit-distance is not guaranteed. For speed, we check if the
+        # matrix is in O(3) and preserves lengths.
+        elif np.all(is_O3(matrix)):  # remain in unit-rep
+            xyz = erfa_ufunc.s2c(self.lon, self.lat)
+            p = erfa_ufunc.rxp(matrix, xyz)
+            lon, lat = erfa_ufunc.c2s(p)
+            rep = self.__class__(lon=lon, lat=lat)
+
+        else:  # route through dimensional representation
+            rep = self._dimensional_representation(
+                lon=self.lon, lat=self.lat, distance=1).transform(matrix)
```

Comment:
p.s. I think it is OK to just attach the differentials here directly, since `SphericalRepresentation` should be able to handle the same differentials....

Reply:
seems to pass the tests.